src/bin/smoosh_tools.rs

```diff
@@ -0,0 +1,787 @@
+
+    jsparagus_repo.ensure_remote(
+        "upstream",
+        "https://github.com/mozilla-spidermonkey/jsparagus.git",
+    )?;
+    let remotes = jsparagus_repo.ls_remote("upstream")?;
+
+    let branch = "refs/heads/ci_generated";
+
+    let ci_generated_sha = match remotes.get(branch) {
+        Some(sha) => sha,
+        None => {
+            return Err(Error::Generic(format!("{} not found in upstream", branch)));
+        }
+    };
+
+    log_info!("ci_generated branch HEAD = {}", ci_generated_sha);
+
+    update_cargo(
+        &mc.smoosh_cargo,
+        OfficialDependencyLineReplacer {
+            sha: ci_generated_sha.clone(),
+        },
+    )?;
+
+    vendor(&mc)?;
+
+    log_info!("Please add updated files and commit them.");
+
+    Ok(())
+}
+
+/// Parse remote string and get GitHub username.
+/// Currently this supports only SSH format.
+fn parse_github_username(remote: String) -> Result<String, Error> {
+    let git_prefix = "git@github.com:";
+    let git_suffix = "/jsparagus.git";
+
+    if remote.starts_with(git_prefix) && remote.ends_with(git_suffix) {
+        return Ok(remote.replace(git_prefix, "").replace(git_suffix, ""));
+    }
+
+    Err(Error::Generic(format!(
+        "Failed to get GitHub username: {}",
+        remote
+    )))
+}
+
+struct BranchInfo {
+    github_user: String,
+    branch: String,
+}
+
+/// Create "generated" branch and push to remote, and returns
+/// GitHub username and branch name.
+fn push_to_gen_branch() -> Result<BranchInfo, Error> {
+    let jsparagus = JsparagusTree::try_new()?;
+
+    let jsparagus_repo = GitRepository::try_new(jsparagus.topsrcdir.clone())?;
+    jsparagus_repo.assert_clean()?;
+
+    log_info!("Getting GitHub username and current branch");
+
+    let origin = jsparagus_repo
+        .get_output(&["remote", "get-url", "origin"])?
+        .trim()
+        .to_string();
+    let github_user = parse_github_username(origin)?;
+
+    let branch = jsparagus_repo.branch()?;
+    if branch == "HEAD" {
+        return Err(Error::Generic(format!(
+            "Detached HEAD is not supported. Please checkout a branch"
+        )));
+    }
+
+    let gen_branch = format!("{}-generated-branch", branch);
+
+    log_info!("Creating {} branch", gen_branch);
+
+    jsparagus_repo.run(&["checkout", "-b", gen_branch.as_str()])?;
+
+    try_finally!({
+        log_info!("Generating files");
+
+        let args = &["all"];
+        let cmdline = format_cmdline("make", args);
+        log_info_cmd!(cmdline);
+        check_command(
+            Command::new("make")
+                .args(args)
+                .current_dir(jsparagus.topsrcdir.clone()),
+            cmdline,
+        )?;
+
+        let gitignore = jsparagus.topsrcdir.join(".gitignore");
+
+
+        log_info!("Modifying .gitignore to not-ignore generated files");
+        let orig_gitignore_content = read_file(&gitignore)?;
+        write_file(&gitignore, orig_gitignore_content.replace("*_generated.rs", ""))?;
```

Comment:
nit: This should not be necessary. `git add` has a `--force` argument to allow adding otherwise ignored files.